### PR TITLE
Identify product issues

### PR DESCRIPTION
**Pull request by perlpunk**
Spike solution which uses the retry job to fetch the results of the other investigate jobs by parsing the investigation comment.

Use the retry job to fetch the results of the other investigate jobs by
parsing the investigation comment.

Issues:
* https://progress.opensuse.org/issues/126527
* https://progress.opensuse.org/issues/109920

Demo:
investigate-identify-product-issues

### CODE REVIEW

File: openqa-investigate [hunk 1]
```diff
@@ -180,18 +180,60 @@ $comment_text"
...
+    local job="$(echo "$job_data" | runjq -r '.job')" || return $?
+
+    local IFS=$'\n'
+    local investigate_comment=($("${client_call[@]}" -X GET jobs/"$origin_job_id"/comments | runjq -r '[.[] | select(.text | contains("Automatic investigation jobs for job") and contains(":investigate:retry**:"))] | sort_by(.id) | first | .text')) || return $?
+
+    for line in "${investigate_comment[@]}"; do
+        if [[ $line =~ :investigate:([^*]+)\*\*:.*\/t([0-9]+) ]]; then
+            investigate_type=${BASH_REMATCH[1]}
+            other_id=${BASH_REMATCH[2]}
+            if [[ $other_id -ne $retry_id ]]; then
```
Comment (Martchus) [inline openqa-investigate:195]:
```suggestion
            if [[ $other_id -ne "$retry_id" ]]; then
```

File: openqa-investigate [hunk 2]
```diff
@@ -180,18 +180,60 @@ $comment_text"
...
+
 post-investigate() {
     local id=$1 old_name=$2
     local rc=0 status
     [[ ! "$old_name" =~ investigate:retry$ ]] && echo "Job is ':investigate:' already, skipping investigation" && return 0
-    result="$(echo "$job_data" | runjq -r '.job.result')" || return $?
+    retry_result="$(echo "$job_data" | runjq -r '.job.result')" || return $?
     investigate_origin="$(echo "$job_data" | runjq -r '.job.settings.OPENQA_INVESTIGATE_ORIGIN')" || return 1
     origin_job_id=${investigate_origin#"$host_url/t"}
+    warn "post-investigate $id $old_name origin_job_id=$origin_job_id"
```
Comment (Martchus) [inline openqa-investigate:220]: Not sure what this warning would tell me if I'd encountered it in a month.

**Review by Martchus — COMMENTED** [review]
I guess this makes generally sense.

File: openqa-investigate [hunk 3]
```diff
@@ -180,18 +180,60 @@ $comment_text"
     fi
 }
 
+fetch-investigate-results() {
+    local origin_job_id=$1 retry_id=$2
+    local state other_job_data state result investigate_type passed
+    local job="$(echo "$job_data" | runjq -r '.job')" || return $?
+
+    local IFS=$'\n'
+    local investigate_comment=($("${client_call[@]}" -X GET jobs/"$origin_job_id"/comments | runjq -r '[.[] | select(.text | contains("Automatic investigation jobs for job") and contains(":investigate:retry**:"))] | sort_by(.id) | first | .text')) || return $?
```
[the remark this answers is not in the document]
Comment (perlpunk) [reply openqa-investigate:189]: > Sorting the whole array is inefficient when you really just want to get the min/max element:

I copied this from your `sync_via_investigation_comment` ;-)

> By the way, I'm wondering whether we really want the comment with the lowest ID here. If we wanted the most recent comment that would be the one with the highest ID (so in my suggestion one would use `max_by`).

Normally we only should have one such investigation comment anyway. So it probably doesn't matter much. Let's suggest `max_by` for the actual ticket then.

File: openqa-investigate [hunk 4]
```diff
@@ -180,18 +180,60 @@ $comment_text"
...
     local id=$1 old_name=$2
     local rc=0 status
     [[ ! "$old_name" =~ investigate:retry$ ]] && echo "Job is ':investigate:' already, skipping investigation" && return 0
-    result="$(echo "$job_data" | runjq -r '.job.result')" || return $?
+    retry_result="$(echo "$job_data" | runjq -r '.job.result')" || return $?
     investigate_origin="$(echo "$job_data" | runjq -r '.job.settings.OPENQA_INVESTIGATE_ORIGIN')" || return 1
     origin_job_id=${investigate_origin#"$host_url/t"}
+    warn "post-investigate $id $old_name origin_job_id=$origin_job_id"
+
+
```
[the remark this answers is not in the document]
Comment (perlpunk) [reply openqa-investigate:222]: That was just debugging

File: openqa-investigate (hunk 2 above)
Comment (perlpunk) [reply openqa-investigate:220]: That was just debugging

File: openqa-investigate (hunk 3 above)
[the remark this answers is not in the document]
Comment (Martchus) [reply openqa-investigate:189]: And I suppose I'm the author of `sync_via_investigation_comment` :-)
So I've created a PR to improve it: https://github.com/os-autoinst/scripts/pull/224

File: openqa-investigate (hunk 1 above)
Comment (perlpunk) [reply openqa-investigate]: Hm, shellcheck doesn't complain here, but I changed it

**Review by okurz — COMMENTED** [review]
+1

**Comment by perlpunk** [comment]
I updated the screenshot with a new comment I just created on osd.

File: openqa-investigate [hunk 5]
```diff
@@ -182,18 +182,82 @@ $comment_text"
...
+    local pass_lgt='' pass_lgb='' pass_lgtb=''
+    product_issue=false
+
+    readarray -t result_lines < <(fetch-investigation-results "$origin_job_id") || return $?
+
+    for line in "${result_lines[@]}"; do
+        if [[ $line =~ ^([^|]+)\|([^|]+)\|([^|]+) ]]; then
+            investigate_type=${BASH_REMATCH[1]}
+            other_id=${BASH_REMATCH[2]}
+            result=${BASH_REMATCH[3]}
```
Comment (Martchus) [inline openqa-investigate:220]: It looks like `other_id` is an unused variable. What about just:
```suggestion
            investigate_type=${BASH_REMATCH[1]} result=${BASH_REMATCH[3]}
```

**Review by Martchus — COMMENTED** [review]
For Bash code it looks quite clean. I only have a few remarks.

File: openqa-investigate [hunk 6]
```diff
@@ -182,18 +182,82 @@ $comment_text"
     fi
 }
 
+fetch-investigation-results() {
+    local origin_job_id=$1
+    local state job investigate_type other_id result investigate_comment comment_lines
+
+    investigate_comment=$("${client_call[@]}" -X GET jobs/"$origin_job_id"/comments | runjq -r '[.[] | select(.text | contains("Automatic investigation jobs for job") and contains(":investigate:retry**:"))] | sort_by(.id) | first') || return $?
```
[the remark this answers is not in the document]
Comment (perlpunk) [reply openqa-investigate]: done, as well as all your other remarks (except one), thanks

File: openqa-investigate [hunk 7]
```diff
@@ -182,18 +182,82 @@ $comment_text"
...
+            result="$(echo "$job" | runjq -r '.result')" || return $?
+            echo "$investigate_type|$other_id|$result"
+        fi
+    done
+}
+
+is-product-issue() {
+    local origin_job_id=$1
+    local state result investigate_type passed result_lines
+    local pass_lgt='' pass_lgb='' pass_lgtb=''
```
[the remark this answers is not in the document]
Comment (Martchus) [reply openqa-investigate:211]: Ah, then let's do it the shellcheck way.

File: openqa-investigate [hunk 8]
```diff
@@ -182,18 +182,77 @@ $comment_text"
...
+
     comment="Investigate retry job: $host_url/t$id"
-    if [[ "$result" == passed || "$result" == softfailed ]]; then
-        comment+=" $result, likely a sporadic failure"
+    if [[ "$retry_result" == passed || "$retry_result" == softfailed ]]; then
+        comment+=" $retry_result, likely a sporadic failure"
     else
         comment+=" failed, likely not a sporadic failure"
+        local product_issue=false
+        is-product-issue "$origin_job_id" || rc=$?
```
Comment (perlpunk) [inline openqa-investigate:249]: Actually I think I should just return `$rc` in every non-zero case here

Comment (perlpunk) [reply openqa-investigate]: done.
And I added a `is-ok` function now.

File: openqa-investigate [hunk 9]
```diff
@@ -182,18 +182,79 @@ $comment_text"
...
-        comment+=" $result, likely a sporadic failure"
+    if is-ok "$retry_result"; then
+        comment+=" $retry_result, likely a sporadic failure"
     else
         comment+=" failed, likely not a sporadic failure"
+        local product_issue=false
+        is-product-issue "$origin_job_id" || return $?
+
+        if "$product_issue"; then
+            comment+="."$'\n'"Test(s) with last good build pass, likely a product issue"
```
Comment (okurz) [inline openqa-investigate:256]: We use the past tense and the actual result for the tests job, I would use the same here, i.e. "passed" or "soft-failed". As alternative use "ok", as we also use within openQA where we define "ok" as "passed or soft-failed"

**Review by okurz — CHANGES_REQUESTED** [review]
Can you explain in more detail why you wait for the retry-comment and extend that instead of it's open comment? If you wanted to reduce the amount of comments and have found a good way to edit existing comments how about always updating the existing initial comment instead?

**Comment by perlpunk** [comment]
> Can you explain in more detail why you wait for the retry-comment and extend that instead of it's open comment?

I don't understand that sentence at all. Can you rephrase? What waiting do you mean?
What do you mean with "it's open comment"?

> If you wanted to reduce the amount of comments and have found a good way to edit existing comments how about always updating the existing initial comment instead?

Yes, that would be possible.

**Comment by okurz** [comment]
> > Can you explain in more detail why you wait for the retry-comment and extend that instead of it's open comment?
> 
> I don't understand that sentence at all. Can you rephrase? What waiting do you mean? What do you mean with "it's open comment"?

Sorry. I meant: Please explain why you add a message to the comment from the "retry" investigation job instead of creating a new comment with just your information about potential product regression

**Comment by perlpunk** [comment]
> Sorry. I meant: Please explain why you add a message to the comment from the "retry" investigation job instead of creating a new comment with just your information about potential product regression

Because I'm doing that at the same time anyway. Posting two different comments didn't even occur to me.

At the time I know that the simple retry test failed, I also investigate the other investigation tests, and then I post a comment with that combined information. I have this information at the same time. It wouldn't make sense to me to just split these two sentences into two comments.

We also don't split the first investigation test list into four separate comments.

Maybe we can clarify this tomorrow in a chat.

File: openqa-investigate (hunk 9 above)
Comment (okurz) [reply openqa-investigate:256]: "Jobs including the last good build are ok, likely a product issue"

**Review by okurz — APPROVED** [review]
